### Notebook 02: Análise de dados - parte 1

Agora que vimos como documentar um notebook, salvar na memória o dataset e dar uma primeira olhada nos dados, é hora de colocar a mão na massa!

Uma análise pode ter duas origens:

- responder a uma pergunta pré-definida;
- não ter pergunta pré-definida alguma e tentar cavar uma pauta.

Confesso que o segundo propósito não é comum, pois quem trabalha com dados geralmente começa uma análise já com uma pergunta formulada na cabeça e busca uma resposta. Mas, por experiência, navegar no meio de números sem propósito também pode render boas histórias. A matéria dos gastos dos vereados de São Paulo com correios nasceu assim, por exemplo.

A verdade é que uma origem não precisa necessariamente eliminar a outra: você pode tentar responder a uma pergunta pré-estabelecida e, ao mesmo tempo, manter a mente aberta para outras coisas que encontrar no caminho.

Para exercitar, bamos subir o primeiro dataset, de despesas dos deputados:

In [1]:
# Importação das ferramentas
import pandas as pd
import numpy as np

# Gravação de dataset de despesas na memória
despesas = pd.read_csv('data/despesas.csv', 
                       encoding = 'latin-1',
                       dtype = {'Deputado': np.str,
                                'Matricula': np.int,
                                'Ano': np.int,
                                'Mes': np.int,
                                'Tipo': np.str,
                                'CNPJ': np.str, 
                                'Fornecedor': np.str,
                                'Valor': np.float})

# Resultado
despesas.shape

(446408, 8)

Com o dataset gravado na memória, cabe ressaltar uma das boas práticas da análise de dados: **não altere o dataset original!**

Por uma questão de lisura, é recomendado deixar o dataset original intacto, sem alteração dos registros.

O que você pode fazer é "batizá-lo" de outras formas. Por exemplo, se eu quero apenas os dados referentes a 2017:

In [2]:
# Recorte de 2017
despesas_2017 = despesas[despesas['Ano'] == 2017]

# Resultado
despesas_2017.shape

(24790, 8)

Ou se eu quero os gastos com a categoria "N - MORADIA":

In [3]:
# Recorte de 2017
despesas_moradia = despesas[despesas['Tipo'] == 'N - MORADIA']

# Resultado
despesas_moradia.shape

(957, 8)

Repare que, tanto no recorte de ano quanto no de tipo, fiz referência ao dataset original -- afinal, ele é a fonte --, mas não o alterei e, por isso, posso retornar a ele sempre que necessário:

In [4]:
# Resultado do dataset original
despesas.shape

(446408, 8)

Vamos ver de maneira mais aprofundada o que temos no dataset original:

In [5]:
# Estatísticas
despesas.describe()

,Matricula,Ano,Mes,Valor
count,446408.000000,446408.000000,446408.000000,446408.000000
mean,300394.687432,2009.099055,6.526742,680.259056
std,106.915937,4.587844,3.331889,1442.157569
min,300010.000000,2002.000000,1.000000,0.000000
25%,300346.000000,2005.000000,4.000000,50.000000
50%,300415.000000,2009.000000,6.000000,127.000000
75%,300468.000000,2013.000000,9.000000,495.042500
max,300595.000000,2018.000000,12.000000,13262.000000


Parece confuso... Mas não é. Juro.

`DataFrame.describe()` mostra algumas estatísticas das colunas que tiverem valores `float` ou `integer` -- ou seja, tudo que for número.

Esta função mostra:

`count`: a contagem de registros na coluna -- no dataset em questão, 446.408 registros ou linhas.<br>
`mean`: a média dos valores de cada coluna -- o que não faz muito sentido para "Matricula", "Ano" ou "Mes", mas faz para "Valor", pois sabemos que o valor médio de todos os registros é 680.26 reais.<br>
`std`: o desvio-padrão, ou seja, a dispersão dos números. O desvio-padrão pequeno indica que os valores estão mais concentrados em torno da média. Exemplo:

`15 + 25 + 20 = 60 => os números estão mais concentrados em torno do meio; logo, desvio_padrão pequeno.`<br>
`05 + 51 + 04 = 60 => os números estão mais dispersos em relação ao meio; logo, desvio-padrão grande.`

`min`: o menor número de cada coluna. No nosso dataset, o menor ano é 2002, e o menor valor é 0 real.<br>
`25%`, `50%` e `75%`: respectivamente o quartil inferior, a mediana e o quartil inferior. Em termos gerais, mostra os valores (ordenados) contidos nos primeiros 25% registros da amostra, na metade da amostra e nos 75% registros da amostra. No nosso dataset, 25% dos valores estão até 50 reais, a mediana é 127 reais, e 75% estão até 495,04 reais.<br>
`max`: o número máximo em cada coluna. No nosso caso, o ano máximo é 2018, e o valor máximo é 13.262,00.

Sabendo isso, a função `DataFrame.describe()` revela, entre outras coisa, que, no nosso dataset:

- há 446.408 registros
- de 2002 a 2018
- com valor médio de 680.26 reais
- sendo o valor mínimo de 0 real
- e o valor máximo de 13.262,00

Mas se ainda parecer confuso -- afinal, não existe um desvio-padrão em mês, ou um valor mínimo em matrícula --, basta selecionar uma coluna única:

In [6]:
# Estatísticas de valores
despesas['Valor'].describe()

count    446408.000000
mean        680.259056
std        1442.157569
min           0.000000
25%          50.000000
50%         127.000000
75%         495.042500
max       13262.000000
Name: Valor, dtype: float64

E quanto aos recortes que fizemos? Vamos ver:

In [7]:
# Estatísticas de valores de 2017
despesas_2017['Valor'].describe()

count    24790.000000
mean       942.263890
std       1718.135442
min          0.000000
25%         98.002500
50%        217.970000
75%        760.000000
max       9334.610000
Name: Valor, dtype: float64

In [8]:
# Estatísticas de valores de moradia
despesas_moradia['Valor'].describe()

count     957.000000
mean     2224.384953
std       946.736637
min         9.220000
25%      1900.000000
50%      2800.000000
75%      2850.000000
max      2850.000000
Name: Valor, dtype: float64

Ok. Temos um recorte de 2017 e outro de moradia. Mas e se quisermos apenas um recorte contemplando os dois critérios? É bastante simples...

Se a lógica de recorte é:

    dataset_secundario = dataset[dataset['coluna'] == 'criterio']

Então, com dois critérios, é:

    dataset_secundario = dataset[(dataset['coluna'] == 'criterio 1') & (dataset['coluna'] == 'criterio 2')]

Vamos testar:

In [9]:
# Recorte com dois critérios
despesas_duplo_recorte = despesas[(despesas['Ano'] == 2017) & (despesas['Tipo'] == 'N - MORADIA')]

# Resultado
despesas_duplo_recorte.head(20)

,Deputado,Matricula,Ano,Mes,Tipo,CNPJ,Fornecedor,Valor
144,ABELARDO CAMARINHA,300257,2017,1,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
149,ABELARDO CAMARINHA,300257,2017,2,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
155,ABELARDO CAMARINHA,300257,2017,3,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
164,ABELARDO CAMARINHA,300257,2017,4,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
173,ABELARDO CAMARINHA,300257,2017,5,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
180,ABELARDO CAMARINHA,300257,2017,6,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
209,ABELARDO CAMARINHA,300257,2017,10,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
217,ABELARDO CAMARINHA,300257,2017,11,N - MORADIA,11354673000166,VARGAS ADM DE BENS E PARTICIPAÇÕES EM SOCIEDAD...,2850.0
2842,ADILSON ROSSI,300432,2017,1,N - MORADIA,05650611602,RONALDO AZZI NOGUEIRA,2800.0
2855,ADILSON ROSSI,300432,2017,2,N - MORADIA,05650611602,RONALDO AZZI NOGUEIRA,2800.0


Funcionou!

Mas a escolha de um recorte vai além disso: em vez de você usar `==` para buscar por valores que atendam exatamente aos critérios estabelecidos, pode usar:

`<`: menor que (em que se exclui o valor)<br>
`<=`: igual ou menor que (em que se inclui o valor)<br>
`>`: maior que (em que se exclui o valor)<br>
`>=`: igual ou maior que (em que se inclui o valor)<br>
`!=`: todos os valores, exceto

Vamos testar. Vamos ver quantos gastos de 2017 com valor superior a 5.000 reais existem:

In [10]:
# Recorte com dois critérios
despesas_2017_acima_5k = despesas[(despesas['Ano'] == 2017) & (despesas['Valor'] > 5000)]

#Resultado
despesas_2017_acima_5k.shape

(1263, 8)

Agora vamos ver em que categorias  estão esses 1.263 registros. Para isso, vamos usar uma função que conta a quantidade de registros:

    DataFrame.value_counts()

In [11]:
# Contagem de registros por tipo
despesas_2017_acima_5k['Tipo'].value_counts()

D - MATERIAIS E SERVIÇOS GRÁFICOS, DE CÓPIAS  E REPRODUÇÃO DE DOCS    535
L - LOCAÇÃO DE BENS IMÓVEIS                                           250
F - SERVIÇOS TÉCNICOS PROFISSIONAIS (CONSULTORIA, PESQUISAS ETC)      164
J - SERVIÇOS DE COMUNICAÇÃO                                           142
K - LOCAÇÃO DE BENS MÓVEIS                                             65
E - MATERIAIS DE ESCRITÓRIO E OUTROS MATERIAIS DE CONSUMO              27
A - COMBUSTÍVEIS E LUBRIFICANTES                                       26
C - MATERIAIS E SERVIÇOS DE MANUT E CONSERV DE VEÍCULOS ; PEDÁGIOS     26
M - MANUTENÇÃO DE BENS MÓVEIS,  IMÓVEIS, CONDOMÍNIOS E OUTROS          19
I - HOSPEDAGEM, ALIMENTAÇÃO E DESPESAS DE LOCOMOÇÃO                     8
H - SERV.UTIL.PÚBLICA (TELEF.MÓVEL/FIXA, ENERGIA, ÁGUA, GÁS ETC)        1
Name: Tipo, dtype: int64

Notamos que a categoria de serviços gráficos contém boa parte das notas acima de 5.000 reais.

Mas quantas notas temos de cada categoria em todo o ano de 2017, descartando o recorte de valor acima de 5.000 reais? 

Bem, já segmentamos a base de ano -- `despesas_2017` -- linhas acima. Vamos usá-la e aplicar a mesma função:

In [12]:
# Contagem de registros por tipo
despesas_2017['Tipo'].value_counts()

A - COMBUSTÍVEIS E LUBRIFICANTES                                      6633
I - HOSPEDAGEM, ALIMENTAÇÃO E DESPESAS DE LOCOMOÇÃO                   3564
H - SERV.UTIL.PÚBLICA (TELEF.MÓVEL/FIXA, ENERGIA, ÁGUA, GÁS ETC)      2953
E - MATERIAIS DE ESCRITÓRIO E OUTROS MATERIAIS DE CONSUMO             2662
M - MANUTENÇÃO DE BENS MÓVEIS,  IMÓVEIS, CONDOMÍNIOS E OUTROS         2068
G - ASSINATURAS DE PERIÓDICOS, PUBLICAÇÕES E INTERNET                 1344
L - LOCAÇÃO DE BENS IMÓVEIS                                           1225
K - LOCAÇÃO DE BENS MÓVEIS                                            1172
C - MATERIAIS E SERVIÇOS DE MANUT E CONSERV DE VEÍCULOS ; PEDÁGIOS    1034
D - MATERIAIS E SERVIÇOS GRÁFICOS, DE CÓPIAS  E REPRODUÇÃO DE DOCS     833
J - SERVIÇOS DE COMUNICAÇÃO                                            627
F - SERVIÇOS TÉCNICOS PROFISSIONAIS (CONSULTORIA, PESQUISAS ETC)       437
N - MORADIA                                                            238
Name: Tipo, dtype: int64

Descobrimos agora que boa parte dos registros são gastos com combustíveis. Quanto será que eles somam?

Para encontrar a resposta, vamos segmentar o dataset `despesas_2017` e usar uma outra função:

    DataFrame.sum()

In [13]:
# Recorte do dataset por tipo
despesas_2017_tanque = despesas_2017[despesas_2017['Tipo'] == 'A - COMBUSTÍVEIS E LUBRIFICANTES']

# Soma dos valores
despesas_2017_tanque['Valor'].sum()

2067121.86

Chegamos a algo interessante: em 2017, os deputados estaduais gastaram 2.067.121,86 com combustível.

Mas esse é o valor total. Qual é a média de gasto com combustível por nota fiscal? Para isso, podemos usar a função `DataFrame.describe()` vista linhas acima, ou ser mais direto, usando uma função que retorna apenas a média:

    DataFrame.mean()

Vamos fazer dos dois jeitos:

In [14]:
# Estatísticas dos valores
despesas_2017_tanque['Valor'].describe()

count    6633.000000
mean      311.642071
std       634.481439
min         0.230000
25%        98.320000
50%       140.090000
75%       261.000000
max      7808.960000
Name: Valor, dtype: float64

In [15]:
# Média dos valores
despesas_2017_tanque['Valor'].mean()

311.64207146087745

Vemos que a média por nota fiscal é 311,64.

Ah, há alguns números além dos dois decimais a que estamos acostumados. Isso porque o Pandas não sabe que estamos trabalhando com valor monetário -- ele é usado em diversas áreas que contemplam mais de duas casas decimais.

Logo você se habitua a automaticamente arrendondar para mais ou para menos, a depender do valor. Mas, se fizer questão de usar apenas dois dígitos, pode recorrer à função:

    DataFrame.round(decimals = 2)

In [16]:
# Arredondamento da média dos valores em dois dígitos
despesas_2017_tanque['Valor'].mean().round(decimals = 2)

311.64

Bem, este notebook contempla muitas coisas. Vale a pena reler e testar tudo o que foi mostrado. No próximo, vamos aprofundar a análise com agrupamento, por exemplo, e descobriremos quais deputados mais gastam.

Vai ser divertido.

Beijos.